shaderObject: Fix potential use of wrong pipeline

qbojj opened the pull request:
closes #269 

It is a quite hacky way to do it as I am not familiar with the code generator.
It breaks the interfaces of `GetShader`/`GetShaderPtr`, but fixes its uses by adding `.pShader`.

What was changed is that every `Shader` holds unique `id` in it and `FullDrawStateData` holds it too, so even if different `Shader` gets the same address as one beforehand it is still differentiated from the old one by `id`.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-ExtensionLayer?pullRequest=270) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-ExtensionLayer?pullRequest=270) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-ExtensionLayer?pullRequest=270) it.</sub>

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build queued with queue ID 4664.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 594 running.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 594 failed.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ziga-lunarg commented:
Thank you for finding and fixing this. One of the most important design goals of this layer is to maintain maximum performance in the most common cases, and while we agree this is a real issue we have some concerns about the architectural and performance impact of this change. If getting a fix merged is time critical for you, could you please see if you can come up with a simpler and/or potentially more elegant solution?

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

qbojj commented:
I've separated the shaders from shader_ids from `FullDrawStateData`.

It reduces overhead on 32-bit due to padding losing only 4 bytes to padding (whole change adds max 44 bytes on all targets and 20 if uint32_t would be enough). It also fixes the `GetShader` and `GetShaderPtr` interfaces.

As for the 64-bit atomic increment on 32-bit targets: it should not be a problem as `ShaderEXT` creation is a low frequency operation.

qbojj commented:
I have done something like this in 831faa7, where `ComparableShader`s equivalent is named `IdShader`. It didn't have a getter, but a public field.

This solution adds more padding to `FullDrawState` struct but I guess maintainability is more important here.

I can modify the code at a later date.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ziga-lunarg commented:
Thank you for these changes, this PR is looking almost ready to merge.

The diff for shader_object_full_draw_state_struct_members.inl looks much larger than it should be. This is probably due to line endings.

Additionally, the private members of ComparableShader don't match the naming convention for other private members in the file. It might also be nice to have a comment about ComparableShader explaining why it exists.

Finally, a small aesthetic question, is there a particular reason why you brought out only pShader on this line? https://github.com/KhronosGroup/Vulkan-ExtensionLayer/blob/75db64887976af1fb7d3ed9d88c55576608fb8b9/layers/shader_object.cpp#L2260-L2261

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build queued with queue ID 13535.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 641 running.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 641 passed.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

qbojj commented:
> The diff for shader_object_full_draw_state_struct_members.inl looks much larger than it should be. This is probably due to line endings.

I do not know what causes that. Line encodings are the same (\n).

>Additionally, the private members of ComparableShader don't match the naming convention for other private members in the file. It might also be nice to have a comment about ComparableShader explaining why it exists.

Fixed and added short explanation.

>Finally, a small aesthetic question, is there a particular reason why you brought out only pShader on this line? 

Just a side-effect of previous commits, I have reverted it.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build queued with queue ID 14240.

ziga-lunarg commented:
The naming convention for private member variables is to add an underscore at the end. Also, the `*` should touch the type rather than the name. For instance, `Shader *shader;` should be `Shader* shader_;` and `Shader *GetShaderPtr()` should be `Shader* GetShaderPtr()`.

The ComparableShader constructor can be defined inline due to how simple it is, this will make it easier to browse the code. Making the constructor a single line like so would work great: `explicit ComparableShader(Shader* shader) : shader_(shader), id_(shader ? shader->id : 0) {}`.

Those are the only things left to do, after those changes are in we should be able to merge. Thank you!

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 647 running.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 647 passed.

qbojj commented:
> The ComparableShader constructor can be defined inline due to how simple it is, this will make it easier to browse the code. 

Inlining constructor cannot be done as it uses implementation of `Shader` (`shader->id`).

Other than that I will fix the rest.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author qbojj not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build queued with queue ID 14352.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 648 running.

ci-tester-lunarg commented:
CI Vulkan-ExtensionLayer build # 648 passed.
